vktrace/src/vktrace_layer/vktrace_lib_trace.cpp

```diff
@@ -136,6 +138,166 @@ void *strip_create_extensions(const void *pNext)
+// small window between the reading of this file and when the dirty bits
+// are cleared in which some other thread might modify mapped memory.
+// We trap these memory writes by setting mapped memory pages to read-only and
+// setting the signal handler for SIGSEGV. The signal handler creates a list
+// of the addresses that caused a SIGSEGV. We keep the list in sighAddrList.
+// We used the semaphore sighAddrListSem to protect access to this list.
+
+static std::list<void *> sighAddrList;
+static vktrace_sem_id sighAddrListSem = NULL;
+
+static void segvHandler(int sig, siginfo_t *si, void *ununsed)
+{
+    size_t pageSize = getpagesize();
+
+    // Note: we use sem_wait and mprotect inside a signal handler.
+    // This is not POSIX compliant, but works on Linux.
+
+    // Add the addr that caused the segv to sighAddrList.
+    // It may not be in a mapped page we are tracking. If it
+    // isn't, we'll detect it later and generate an error then.
+    vktrace_sem_wait(sighAddrListSem);
+    sighAddrList.emplace_front(si->si_addr);
+    vktrace_sem_post(sighAddrListSem);
+
+    // Change protection of this page to allow the write to proceed
+    mprotect((void *)((uint64_t)si->si_addr & ~(pageSize-1)), pageSize, PROT_READ|PROT_WRITE);
+}
+
+
+// This function is called when we need to update our list of mapped memory
+// that is dirty.  On Linux, we use the /proc/self/pagemap to detect which pages
+// changed. But we also use mprotect with a signal handler for the rare case
+// in which mapped memory is written to between when we read the dirty bits
+// from /proc/self/pagemap dirty and clear the dirty bits.
+
+void getMappedDirtyPagesLinux(void)
+{
+    LPPageGuardMappedMemory pMappedMem;
+    PBYTE addr, alignedAddrStart, alignedAddrEnd;
+    uint64_t nPages;
+    VkDeviceMemory mappedMemory;
+    off_t pmOffset;
+    int64_t index;
+    size_t pageSize = getpagesize();
+    size_t readLen;
+    VKAllocInfo *pEntry;
+    struct sigaction sigAction;
+    static int pmFd = -1;
+    static std::vector<uint64_t> pageEntries;
+    
+    // If pageguard isn't enabled, we don't need to do anythhing
+    if (!getPageGuardEnableFlag())
+        return;
+
+    // Open pagefile, initialize sighAddrList semaphore, and set the SIGSEGV signal handler
+    // Also clear all dirty bits because some older kernels may require it.
+    if (pmFd == -1)
+    {
+        pmFd = open("/proc/self/pagemap", O_RDONLY);
+        if (pmFd < 0)
+            VKTRACE_FATAL_ERROR("Failed to open pagemap file.");
+
+        if (!vktrace_sem_create(&sighAddrListSem, 1))
+            VKTRACE_FATAL_ERROR("Failed to create sighAddrListSem.");
+
+        sigAction.sa_sigaction = segvHandler;
+        sigfillset(&sigAction.sa_mask);
+        sigAction.sa_flags = SA_SIGINFO;
+        if (0 != sigaction(SIGSEGV, &sigAction, NULL))
+            VKTRACE_FATAL_ERROR("sigaction sys call failed.");
+        getPageGuardControlInstance().pageRefsDirtyClear();
```

Comment:
As soon as I submitted this revision, I realized the call to pageRefsDirtyClear should not be done here. It should be moved to the top of  __HOOKED__vkMapMemory.